File: python/detect_squares.py
from collections import defaultdict
class DetectSquares(object):

    def __init__(self):
        self.freqMap = defaultdict(int)
        self.points = []
        
    # O(1) time | O(1) space
    def add(self, point):
        """
        :type point: List[int]
        :rtype: None
        """
        self.freqMap[(point[0], point[1])] += 1
        self.points.append(point)
        

    # O(n) time | O(1) space
    def count(self, point):
        """
        :type point: List[int]
        :rtype: int
        """
        x = point[0]
        y = point[1]
        res = 0

        for pt in self.points:
            px = pt[0]
            py = pt[1]

            if abs(x - px) != abs(y - py) or x == px or y == py:
                continue

            res += self.freqMap[(x, py)] * self.freqMap[(px, y)]

        return res
        

File: src/main/java/september_2025/DetectSquares.java
package september_2025;

import java.util.*;

public class DetectSquares {

    public static void main(String[] args) {
        DetectSquares detectSquares = new DetectSquares();
        detectSquares.add(new int[] {3, 10});
        detectSquares.add(new int[] {11, 2});
        detectSquares.add(new int[] {3, 2});

        int count = detectSquares.count(new int[]{11, 10});
        System.out.println(count);

        int count1 = detectSquares.count(new int[]{14, 8});
        System.out.println(count1);

        detectSquares.add(new int[] {11, 2});

        int count2 = detectSquares.count(new int[]{11, 10});
        System.out.println(count2);
    }

    Map<List<Integer>, Integer> countsMap;
    List<int[]> points;
    public DetectSquares() {
        countsMap = new HashMap<>();
        points = new ArrayList<>();
    }

    // O(1) time
    public void add(int[] point) {
        countsMap.put(Arrays.asList(point[0], point[1]), countsMap.getOrDefault(Arrays.asList(point[0], point[1]), 0) + 1);
        points.add(point);
    }

    // O(n) time | O(n) space
    public int count(int[] point) {
        int res = 0;
        int x = point[0];
        int y = point[1];

        for (int[] p : points) {
            int px = p[0];
            int py = p[1];

            if (Math.abs(x - px) != Math.abs(y - py) || x == px || y == py) {
                continue;
            }

            res += countsMap.getOrDefault(Arrays.asList(x, py), 0) * countsMap.getOrDefault(Arrays.asList(px, y), 0);
        }

        return res;
    }

}
